Cardano.Api -> Cardano.Api.Typed and HardForkBlock

rvl opened the pull request:
### Issue Number

Relates to #1870.

### Overview

- Updates cardano-node to latest master version.
- Updates libraries to version from cardano-node master.
- Cardano.Api → Cardano.Api.Typed changes.
- Change `ShelleyBlock` → `CardanoBlock sc` in "Compatibility" module
- Change `ShelleyBlock` → `CardanoBlock sc` in network module
- Add in byron implementations of conversion functions for two-era mode.
- Add CLI support for two-era mode.
- Use LocalStateQuery to get the history Interpreter from the node.
- Update shelley transaction layer to use new extended keys support in Cardano.Api.Typed.

### Comments

- [x] Fix all the type errors in packages other than cardano-wallet-shelley.
- [x] Transaction layer works during shelley era
- [ ] Transaction layer works during byron era
- [x] Manual testing against [mainnet_candidate](https://hydra.iohk.io/job/Cardano/iohk-nix/cardano-deployment/latest/download/1/index.html) which is in byron era -- can sync and restore empty wallet
- [x] Manual testing against [shelley_qa](https://hydra.iohk.io/job/Cardano/iohk-nix/cardano-deployment/latest/download/1/index.html) which is in shelley era -- can sync and restore empty wallet
- [ ] Unit tests passing
- [x] Merge integration test cluster setup from #1924, and fix.

rvl committed:
Bump cardano-node to latest master

rvl committed:
Cardano.Api → Cardano.Api.Typed and HardForkBlock support

Anviking committed:
Add support for Byron;Shelley integration test cluster

- Adjusted configs to work with the new version of cardano-node
- withCluster now takes three actions:
  - onByron
  - onFork (when in shelley)
  - onClusterStart (when the stake pools have fully started)
- CLI faucets are now byron facuets, and are generated automatically.
This is both easier and more elegant.
- The underlying logic of withCluster was kept abstract for easier
experimentation.

Contains a few squashed commits from Rodney.

rvl committed:
Regenerate nix

rvl committed:
Bump iohk-nix

rvl committed:
PostTx type family

Anviking commented on lib/shelley/src/Cardano/Wallet/Shelley/Transaction.hs:
```diff
@@ -192,23 +197,30 @@ mkTx
+                     | null wdrls = []
+                     | otherwise = [_ (getRawKey rewardAcnt, pwdAcnt)]
+
+            tx <- signShelleyTransaction unsigned (paymentKey:stakeKey)
```
This seems weird to me, what was the problem here?

Anviking replied:
Also, it seems you almost went in the direction of creating a byron vs Shelley tx based on the key/witness tag, but we may spend from byron addresses in shelley.

 So we need to either on the era we're in (or if there's something in cardano-api that does that automatically)?

rvl replied:
Basically what happened is there were functions removed that we were using, so I attempted to use the new API and broke it.

rvl committed:
Revert Cardano.Wallet.Shelley.Transaction back to master, start again.

rvl committed:
bump iohk-nix

rvl committed:
Debugging

rvl committed:
Bump cardano-node to latest master

Anviking committed:
Start converting Shelley TransactionLayer to new Cardano.Api.Typed

Anviking committed:
Add toCardanoNetworkId and remove @n from tl

rvl committed:
Fix cli genesis

rvl committed:
Fix shelley genesis file in $CARDANO_NODE_CONFIGS

It has to be exactly the same, because the file hash is used as the
shelley nonce.

rvl committed:
Logging more info

Anviking committed:
Fill in a few more gaps in Shelley tl

rvl committed:
Update shelley transactionspec

rvl committed:
TransactionSpec: replace type errors with runtime errors

rvl committed:
Test compile fixes

Anviking committed:
Re-add construction of dummy byron witnesses

Anviking committed:
Remove redundant imports

Anviking committed:
Add HasNetworkId class replacing toCardanoNetworkId

rvl committed:
Build fixes

Anviking committed:
Fix FIXMEs in unit tests, re-add shelley toSealed

rvl committed:
Remove shelley genesis CLI options - only use byron genesis

Anviking committed:
Update (slightly) failing fee estimation goldens

It seems the fee estimation has changed, and these tests just need to be
adjusted slightly.

rvl committed:
Fix compile

rvl committed:
toShelleyGenTx -> unsealShelleyTx

rvl committed:
Rename

rvl committed:
Revert "PostTx type family"

This reverts commit 2d39731592713297fec474cf2dc89d5c0a6de3c1.

rvl committed:
fixup

rvl committed:
Fix toSealed

rvl committed:
Include byron address attributes in tx witnesses

KtorZ commented on lib/core/src/Cardano/Wallet/Network.hs:
```diff
@@ -141,18 +143,20 @@ data NetworkLayer m target block = NetworkLayer
 instance Functor m => Functor (NetworkLayer m target) where
     fmap f nl = nl
         { nextBlocks = fmap (fmap f) . nextBlocks nl
+        , stakeDistribution = error "fixme: functor instance"
```
:point_up: Just leaving a marker here.

KtorZ commented on lib/shelley/src/Cardano/Wallet/Shelley/Transaction.hs:
```diff
@@ -416,39 +383,45 @@ computeTxSize proxy pm witTag action cs =
+        dummyPoolId :: Cardano.PoolId
+        dummyPoolId = fromMaybe (error "dummyPoolId couldn't be constructed")
+            $ Cardano.deserialiseFromRawBytes (Cardano.AsHash Cardano.AsStakePoolKey)
+            $ BS.pack $ replicate 32 0
```
This needs to be 28 bytes.

KtorZ commented on lib/shelley/src/Cardano/Wallet/Shelley/Transaction.hs:
```diff
@@ -416,39 +383,45 @@ computeTxSize proxy pm witTag action cs =
-            (if Map.null wdrls then Set.empty else Set.singleton (dummyWitness "0"))
+        (
+            map dummyWitnessUniq (fst <$> CS.inputs cs)
+            <> [dummyWitness "0" | null wdrls]
```
```hs
>>> [14|True]
[14]

>>> [14|False]
[]
```

This creates a dummy witness if the witness set is empty. Which is, the opposite of what we want :grimacing: ... The list comprehension is shorter / somewhat more elegant but also more confusing, so I'd rather put back the good old `if null then [] else [...]` which is less error-prone.

Anviking replied:
Sorry, hlint asked for it, and thought: fine.

KtorZ replied:
hlint is stupid. We are not (arguably, but I will argue anytime). Sometimes, when the suggestion isn't sound, we might as well turn it off. I reverted this to a more readable / less error-prone version.

KtorZ commented on lib/shelley/src/Cardano/Wallet/Shelley/Transaction.hs:
```diff
@@ -506,11 +483,13 @@ computeTxSize proxy pm witTag action cs =
     bloatChaff n = BL.toStrict . BL.take (fromIntegral n) . BL.cycle
 
+    -- TODO: Surely we can allow byron witnesses paying for certificates?
+    -- Should be no reason to case here.
```
Actually we do not. We don't have hybrid wallets, so a wallet that would require Byron witnesses is by construction a "Byron wallet" with no delegation / rewards capability.

Anviking replied:
To me it seems like there is not benefit to not allowing it, and limits our options for e.g. integration test setup.

KtorZ replied:
The benefit is lowering complexity. There are a lot of things that hold together precisely because only Shelley wallets are able to delegate and have withdrawals. Changing this would add a great deal of complexity AND, it'll remove the incentive there's for people to move to Shelley wallets (remember that Byron wallets have security flaws).

Anviking replied:
Changing `mkTx` should have no user-facing effect since shelley wallets' won't select byron inputs.

I also suspect the code here would be more elegant, and we would have the option to use a slightly more flexible `mkTx` in integration tests / for scripts.

KtorZ replied:
From within the transaction layer it shouldn't really have any visible impact indeed. And it's perhaps best to not make any assumption about the "outside context" within the transaction layer actually.

KtorZ commented on lib/shelley/src/Cardano/Wallet/Shelley/Transaction.hs:
```diff
@@ -522,87 +501,114 @@ lookupPrivateKey keyFrom addr =
-    addrAttr = Byron.mkAttributes $ Byron.AddrAttributes
-        (toHDPayloadAddress addr)
-        (toByronNetworkMagic protocolMagic)
+    sk = CC.SigningKey $ Crypto.HD.xPrvChangePass pwd BS.empty prv
```
Would be nice to be consistent with the above. I'll refactor.

KtorZ committed:
fix stylish-haskell issues in Primitive.Slotting

KtorZ committed:
slightly refactor 'mkByronWitness' to be consistent with 'mkShelleyWitness'

The key unencryption is common to both, so we might as well call things the same way

KtorZ committed:
fix fee estimation for byron;shelley transaction layer

There were two issues:

- 32 byte hashes were used for pool ids
- withdrawal witnesses were added when not needed, and vice-versa

KtorZ committed:
revert adding 'block' type parameter to GetStakeDistribution type family

Instead, I've changed the signature to take a wallet 'BlockHeader' instead of a cardano-noed Point, such that the conversion is done within the implementation of the network layer, instead of being done by the caller.

KtorZ committed:
fix staging configuration parsing for cardano-wallet-shelley

KtorZ commented on lib/shelley/src/Cardano/Wallet/Shelley/Transaction.hs:
```diff
@@ -522,87 +506,117 @@ lookupPrivateKey keyFrom addr =
-    SL.makeBootstrapWitness txHash signingKey addrAttr
+    -> Cardano.Witness Cardano.Shelley
+mkByronWitness (Cardano.ShelleyTxBody txbody _) nw addr encryptedKey =
+    Cardano.ShelleyBootstrapWitness $
```
Any reason why we refactored all these actually and no longer use `makeBootstrapWitness` :thinking: ? This code was quite tricky to get right in the first place, so I am still very doubtful about it working correctly now.

KtorZ replied:
I ended up changing it back in 59535f54c3d49d98c20d432f19dc60c8f1eef172

rvl replied:
Yes it was confusion related to `Cardano.Api.Typed.makeShelleyBootstrapWitness`.

KtorZ committed:
revert shelley bootstrap witness creation, re-use makeBootstrapWitness

This was changed perhaps because of lack of documentation / context on cardano-api. Yet, since this was already
quite a tricky part to get right, I'd rather stick with simpler code that we know worked well.

Anviking committed:
Print cartouche when waiting for hard-fork to finish

Anviking committed:
wip: add working sendFaucetFundsTo

Anviking committed:
Remove shelley wallets from shelley genesis (now automatic)

Anviking committed:
wip: Add byron faucet funds

Anviking committed:
Fixup: Add NO_POOLS option, and remove dev functions

KtorZ committed:
some cosmetic changes while reviewing code

Anviking committed:
fixup Crypto.HD

Anviking committed:
Fixup: remove submitTxNoRetry

KtorZ committed:
remove multi-mode support (default to byron;shelley) and rename setup files for consistency

KtorZ committed:
replace hard-coded occurences of 'Secure Passphrase' with a named constant

And in particular, fix the inconsistency between fixture wallets and
empty wallets that have different passphrases!

Anviking committed:
Automate generation of shelley faucet funds

Anviking committed:
Re-add call to prepareKeyRegistration, fix mistake in addr

Anviking committed:
Adjust documentation in shelley-genesis

KtorZ committed:
add little correction to size estimation for Byron change addresses

  indeed, for the size estimation, we generate dummy change outputs that necessary are shelley addresses. Yet, Byron addresses have different sizes, and in particular, Byron are longer than shelley addresses (Icarus are however shorter). So, we apply a correction after the whole size calculation for each change output because generating dummy Byron addresses would be a real pain in the a**

KtorZ committed:
get rid of Shelley's fee estimator from test integration

  This has been heavily reliable and is actually duplicating logic compared to the API. So instead, we query directly the API for fee estimation within the scenario and rely on a single implementation of fee estimation. This commits break every test in Shelley that are using the fee estimator, they'll be updated in subsequent commits

KtorZ committed:
replace fee estimator from Scenario/API/Shelley/Transactions with API calls

  I also removed some scenarios that were completely redundant with others (testing that the fee estimation endpoint works correctly is intrinsically done in other scenarios now that are calling it for pretty much every transactions).

KtorZ committed:
fix 'slotRangeFromTimeRange' time interpreter query

The lifting was not really done wrong. We shouldn't return 'Nothing' if one of the given bound is 'Nothing'.

rvl committed:
Use genesis block parameters for waitForHardFork in tests

KtorZ committed:
fix race-condition in big wallet migration scenario + temporarily disable it

  The original balance is okay now, but somehow, there are less transactions than expected that are created...

rvl committed:
Fix hard for sleep delay

Anviking committed:
Make sure to use the correct timeInterpreter in DB

Anviking committed:
Simplify confusing TestShelleyHardFork setup (still trigger manually)

Anviking committed:
Fixup unused imports in slotting

Anviking committed:
Use SlotNo without qualified Cardano prefix in Slotting module

KtorZ committed:
get rid of #feeEstimator in shelley specs for stake pools

  Most scenarios are now passing, but scenarios related to rewards are still failing.

KtorZ committed:
remove invalid function 'toCardanoStakeAddress'

  The ChimericAccount in cardano-wallet corresponds to a key hash. In order to turn it into a stake address, we require more work and in particular, need a network discriminant.
  This function was unused so it didn't cause any arm, yet.

KtorZ committed:
bump to cardano-1.17.0

KtorZ committed:
increase maxLovelaceSupply in shelley to have a non-null reserve (& rewards)

  We did increase the amount of the launcher faucets that setup the various shelley faucets by 3 order of magnitude. This resulted in the circulation of ada being larger than the specified max supply! As a consequence, no more rewards were distributed at all. Solving this the easy way by increasing the supply.

KtorZ committed:
reduce 'eventually' timeout upper bound to a more reasonable value

KtorZ committed:
use time interpreter in the network information endpoint instead of the legacy API

KtorZ committed:
compare epoch numbers in the shelley stake pools specifications instead of full timestamp

  The main reason is because, in the test scenario we do not have a full time interpreter so it becomes quite complex to estimate the precise time. The epoch however should be fairly easy to identify.

KtorZ committed:
fix stylish-haskell & hlint & weeder

Commit:
fixup! Regenerate nix

KtorZ committed:
use BYRON genesis for initial cluster funds

  Indeed, there are no initial funds in the Shelley genesis. The current state of affairs was doing really weird/wrong things but we didn't quite noticed because so far, we haven't enabled a single test in the shelley scenarios that uses Byron funds. Yet, we still need to be careful about parsing the initial key->value map of initial funds for sequential wallets

paweljakubas committed:
make feeOpt dependent on minimum utxo

KtorZ committed:
add more debug & info logs during for more visibility during migration

  As users / monitorers, we were quite blind to what exactly happens during a migration. Nothing was logged beyond the API calls (which don't have much information in INFO). Now we get proper log with the distribution before and after migration, as well as, in DEBUG, the complete migration plan. That's better for debbugging :)

KtorZ committed:
nuke blatant lurking bug in the migration algorithm... sitting there for months

  So, what this little thing did was to interrupt the ENTIRE migration if one selection in the batch ended up being full of dust. That's wrong, because that might just be bad luck and there could be (as it is right now in integration tests) many coins to still process!

KtorZ committed:
add 'leftovers' to api migration information

  This gives slightly more information about the final amount that one can expect on a resulting wallet. The leftovers are those little dust coins that haven't been included in the migration because they were judge too expensive to have. This means, that in practices, we have: new balance = original balance - migration cost - leftovers, and after migration, the balance of the original wallet is exactly 'leftovers'.

KtorZ committed:
re-enable 'big wallet' migration test, now that everything is fixed.

KtorZ committed:
add non-null dust & property reflecting it for 'rebalanceSelection'

  This fails now, since 'rebalanceSelection' completely disregard any
  form of dust threshold.

KtorZ committed:
acknowledge dust threshold in fee balancing

  I greatly simplified the function as well, it probably handles slightly worse some edge-cases but it's probably for the best in terms of maintenance burden.

KtorZ committed:
tweak fee estimation on migrations so that it assumes worst case scenario for addresses.

  See note in code.

KtorZ committed:
align Byron migration scenarios with recent changes done on their Shelley equivalent (leftovers)

KtorZ committed:
turn on integration scenarios using Byron wallets (and requiring bootstrap witnesses)

KtorZ committed:
revert refactor on 'rebalanceSelection' and coalesce dust AFTER rebalancing

  What we loose from this might be exact balancing in some cases (when the transaction is exactly balanced but coalescing dust may remove one or more change output and change the fee).
  So we might hit problems with some edge-cases on Jörmungandr but things should be fine on cardano-node where we might occasionally pay a little bit more than necessary...

KtorZ committed:
adjust pool pool retirement configuration and retire first pool when it's safe

  Retiring at epoch 1 causes some race condition with the cluster setup.
  Indeed, we now know to wait for the hard fork to begin and it
  sometimes happen that it's already too late to submit the certificate
  for a retirement in epoch 1. Therefore, retirement now happens a bit
  later. Also, I've bumped the retirement at epoch '1_000' to '100_000'
  because we are actually reaching epoch 1_000 pretty early in the test
  suite and even up to 10k epochs!

KtorZ committed:
Merge pull request #1947 from input-output-hk/paweljakubas/proper-treatment-of-dusting-threshold

make feeOpt dependent on minimum utxo

KtorZ committed:
use correct pool id instead of arbitrary one and fix new file names from 1.17.0
